### ISSUE TITLE
Update modes

### ISSUE THREAD

**Issue by Korijn** [pygfx/pygfx#495]
Somewhat related to discussion in #490

I was able to pool the collective memory of HN on this topic: https://news.ycombinator.com/item?id=35435669

What really struck me as an interesting option is the way Bevy handles it: https://docs.rs/bevy/0.10.1/bevy/winit/enum.UpdateMode.html

But we should also remain attentive to the "reactive" methodology where a central state object is used to trigger redraws (as @berendkleinhaneveld has implemented in his project), and this method is not supported by Bevy either.

**Comment by almarklein** [comment]
We are already close to the reactive mode (I think we're actually closer to `ReactiveLowPower`). The `request_draw` mechanism that we have is pretty neat in that it is more precise than e.g. drawing whenever an event happens (most button-less mouse move events don't require a redraw). It also works well for reactive rendering.

I do like the idea to queue events, and process them all right before a draw. We already rate-limit events, but bringing it under the update step together with drawing is even faster, and I suspect also leads to simpler code.

How this could work:

* Schedule updates at a fixed interval using the GUI's scheduler (e.g. at 30fps, 60 fps, or as fast as possible).
* On each update:
  * Process all events in the queue.
  * Emit a custom "update event" that animations can use. (we currently call this before_draw).
  * Both of these can request new draws.
  * Only when a new draw has indeed been requested, we do the actual draw.

This should also work nice with reactive rendering:
- Disable pygfx from calling `request_draw` by itself.
	- Setting `auto_update=False` on the controller.
	- Could even subclass the renderer to make `request_draw()` do nothing.
- Listen to events, and  use them to update the app's state object instead of changing the state of pygfx world objects.
- React to changes in the app state, two options:
	- 1: Watch state changes as they occur, and when the state changes, request a new draw.
	- 2: Check for state changes in the 'update' event, and request a new draw when state has changed.

In order to implement this, we 'd need the following changes:
* in wgpu:
  * implement an update-loop that runs at a steady interval, as outlined above.
* in pygfx:
  * replace the before_draw for the update event.

**Comment by Korijn** [comment]
>   * Emit a custom "update event" that animations can use. (we currently call this "before_render").

I assume this would be emitted by the GUI layer, right?

**Comment by almarklein** [comment]
Yes, I think the canvas would emit it, and the renderer would re-emit it just like it does for the other events.

The main question is how to let animations (e.g. controller smoothing) do updates. The above is one solution. There might be others. The only other solution I could think of is to introduce (in wgpu-py) a `Timer` object, that can be subclassed, and that gets "ticked" periodically by the event loop. There are some open ends like it assumes there is exactly one loop, and what happens if multiple canvases are at play.

**Comment by Korijn** [comment]
> The main question is how to let animations (e.g. controller smoothing) do updates. The above is one solution.

I really like your proposal because it means the renderer and controller remain unaware of the event loop, and you can still pick them up and apply them independently to other problems.

> * Schedule updates at a fixed interval using the GUI's scheduler (e.g. at 30fps, 60 fps, or as fast as possible).

I don't think it's safe to assume the interval will actually be fixed in practise. If drawing a single frame takes too long it won't be achievable. So let's call it a "target" update rate?

>   * Process all events in the queue.

Aggregated / with the buffering approach? Or not?

>   * Only when a new draw has indeed been requested, we do the actual draw.

Would it still be trivial/easy for users to request a draw every frame?

**Comment by almarklein** [comment]
> I don't think it's safe to assume the interval will actually be fixed in practise. If drawing a single frame takes too long it won't be achievable. So let's call it a "target" update rate?

Good point!

> Aggregated / with the buffering approach? Or not?

Yes, we can use the same aggregation method as we do now (e.g. subsequent mouse move events can be combined).

> Would it still be trivial/easy for users to request a draw every frame?

Yes, I guess that would work the same as it does now. Though if this system works well, it wouldn't be needed.

**Comment by Korijn** [comment]
We have seen the following examples in our current codebase that would need an update event:

* Controllers
* Gizmo
* Helpers (light helpers are an example, but the other helpers could also use it)

**Comment by almarklein** [comment]
Stricly speaking the Gizmo is a helper. Though one that already needs events for its interaction.

**Comment by almarklein** [comment]
Nice read related to this topic: http://gameprogrammingpatterns.com/game-loop.html

**Comment by Korijn** [comment]
I think what we should do here is (in order):

* Define the update modes we want to support
* Define the API requirements and design for the GUI layer to support all of the defined update modes
* Define the implementation considerations for interactive pygfx elements like controllers, Input, Time, animations, etc

**Comment by Korijn** [comment]
# Update modes

* Unlimited/continuous
  * [ ] Verify that indeed it is possible to render at unlimited FPS and it is not still limited by some factor (e.g. vsync, max FPS, GUI layer... anything)
* Target FPS (limited fps)
* Manual/on-demand (only when user requests a frame to be drawn)

These are all already supported.

# Options

* Vsync (canvas option)
* Present mode (depends on vsync choice)
* Present mode options (e.g. FIFO queue size) > depends on wgpu-native to expose this
  * [ ] Check if we can expose such options?
* Low power mode (don't render when the screen is hidden etc, this is supported at GUI level and pygfx is not really concerned with it)

# Architectural decisions

* Interactive elements of pygfx should be separated from the renderer.
  * GUI layer
  * Events
  * Controllers
  * Gizmo

The reasoning is that interactive elements require the user to adopt a certain paradigm (event based programming in this case) which may not be compatible with the user's software architecture.

* Users should always be able to use the renderer independently from the GUI layer and interactive elements.

# Loops

* Closed loop (still callback based)
* Event driven
* State change driven

Are all supported already with the current GUI layer. We could add examples to showcase each.

Alternatively users can bring their own GUI layer and implement a true closed loop with their own event system.

# Simulations, animations, ticks

We can implement a `canvas.tick` callback to implement a fixedUpdate style paradigm.

This can be implemented on top of the existing API and therefore can be postponed till after 1.0.0.

We can provide this as an example initially and consider integrating it into API independently.

# Event aggregation

We aggregate events to minimize event handling overhead.

However these are flushed on an independent timer.

This system can be improved:

* Flush events in tandem with animate calls.

**Comment by almarklein** [comment]
After reading this: I think having a page in the docs on this topic would be very valuable. We can explain the different options and possibilities, talk about best practices, refer to some of the links above, and to some of our examples, etc.

**Comment by Korijn** [comment]
Brainfart:

The vision is that the renderer should be just that: a renderer. Events and wobject lifecycles should be implemented outside of the renderer.

I'm thinking of implementing some kind of EventLoop class hierarchy that actually calls all the different lifecycle methods and triggers events. Different implementations can offer different lifecycle models, e.g. ondemand, closed and event based being the primary models I can foresee

**Comment by kushalkolar** [comment]
> Brainfart:
> 
> The vision is that the renderer should be just that: a renderer. Events and wobject lifecycles should be implemented outside of the renderer.
> 
> I'm thinking of implementing some kind of EventLoop class hierarchy that actually calls all the different lifecycle methods and triggers events and calls the renderer

Do you think this could also help with clearing events attached to world objects, and help simplify their garbage collection?

**Comment by almarklein** [comment]
What do you mean with "wobject lifecycles"? Do you mean the wgpu objects for a wobject (i.e. `pipeline.py`)?

**Comment by almarklein** [comment]
This probably needs more thought and discussion, but for now I think I feel what you mean.

**Comment by almarklein** [comment]
Korijn and I spent a day last week about this broad topic. We've the broad outlines of a plan. I updated the top post and also formulated some more concrete tasks. This will be a lot of work, and the API will change quite a bit (all examples will likely be updated). This will be fun! 🚀

**Comment by Korijn** [comment]
Great overview!

**Comment by Korijn** [comment]
I think the important thing for me is that pygfx is a rendering library, not a gui framework. It feels like we are very prone to feature creep.

We can have all the gui stuff, but let's ensure it's cleanly separated from the core.

**Comment by almarklein** [comment]
There's a lot of technical details here. Most people other than @Korijn and me just want to know what all this means for their code :) I created #874 for this purpose.
